File: 0198-house-robber/0198-house-robber.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        
        n = len(nums)
        
        if n == 1:
            return nums[0]
        if n == 2:
            return max(nums[0], nums[1])
        if n == 3:
            return max(nums[0]+nums[2], nums[1])
        
        
        dp = [0]*n
        dp[n-1] = nums[n-1]
        dp[n-2] = nums[n-2]
        dp[n-3] = nums[n-3] + nums[n-1]
        
        for i in range(n-4, -1, -1):
            dp[i] = nums[i] + max(dp[i+2], dp[i+3])
        
        return max(dp[0], dp[1])

                

File: 0198-house-robber/0198-house-robber.cpp
class Solution {
public:
    int rob(vector<int>& nums) {

        const int n = nums.size();

        int dp_1 = 0;
        int dp_2 = 0;
        int temp = 0;
        for (auto num:nums){
            temp = max(num + dp_2, dp_1);
            dp_2 = dp_1;
            dp_1 = temp;
            
        }
        
        return dp_1;

    }
};